File: C++/Validate-Binary-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int prev;
    bool first;
public:
    bool helper(TreeNode* root) {
        if (root == NULL) return true;
        if (!helper(root->left)) return false;
        if (first == false && root->val <= prev) return false;
        first = false, prev = root->val;
        if (!helper(root->right)) return false;
        return true;
    }
    bool isValidBST(TreeNode* root) {
        first = true;
        prev = INT_MIN;
        return helper(root);
    }
};

File: Python/Validate-Binary-Search-Tree.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def inorder(self, root, seq):
        if root == None:
            return
        self.inorder(root.left, seq)
        seq.append(root.val)
        self.inorder(root.right, seq)
    def isValidBST(self, root):
        seq = []
        self.inorder(root, seq)
        length = len(seq)
        for i in range(1, length):
            if seq[i] <= seq[i-1]:
                return False
        return True